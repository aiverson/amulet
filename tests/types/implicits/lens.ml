external val tostring : 'a -> string = "tostring"
external val writeln : string -> unit = "print"

type identity 'a = Identity of 'a
let runIdentity (Identity x) = x

type const 'a 'b = Const of 'a
let getConst (Const x) = x

type functor 'f = Functor of (forall 'a 'b. ('a -> 'b) -> 'f 'a -> 'f 'b)
let (<$>) ?(Functor f) = f

type applicative 'f = Applicative of (forall 'a 'b. 'f ('a -> 'b) -> 'f 'a -> 'f 'b) * (forall 'a. 'a -> 'f 'a)
let (<*>) ?(Applicative (f, _)) = f
let pure ?(Applicative (_, f)) = f

type monoid 'm = Monoid of 'm * ('m -> 'm -> 'm)
let mappend ?(Monoid (_, f)) = f
let mempty ?(Monoid (m, _)) = m

type show 'a = Show of 'a -> string

let show ?(Show f) = f
let print ?(Show f) x = writeln (f x)

let implicit show_string = Show (fun x -> x)
let implicit show_int = Show (fun (x : int) -> tostring x)
let implicit show_bool = Show (function | true -> "true" | false -> "false")
let implicit show_pair ?(sa : show 'a) ?(sb : show 'b) : show ('a * 'b) =
  Show (fun (a, b) -> "(" ^ show a ^ ", " ^ show b ^ ")")

let implicit functor_identity =
  Functor (fun f (Identity x) -> Identity (f x))

let implicit functor_const =
  Functor (fun _ (Const x) -> Const x)

let implicit applicative_identity =
  Applicative (fun (Identity f) (Identity x) -> Identity (f x), Identity)

let implicit applicative_const ?(Monoid (mempty, mappend) : monoid 'x) =
  Applicative (fun (Const (f : 'x)) (Const x) -> Const (mappend f x), fun _ -> Const mempty)

let implicit functor_from_applicative ?(ap : applicative 'f) = Functor (fun f x -> (pure f) <*> x)

let f & g = fun x -> f (g x)
let x |> f = f x

let fst ?f k (a, b) = (,b) <$> k a
let snd ?f k (a, b) = (a,) <$> k b

let l .~ b = runIdentity & l (fun _ -> Identity b)
let s ^. l = getConst (l Const s)

type list 'a = Nil | Cons of 'a * list 'a
let (::) x y = Cons (x, y)

let implicit show_list ?(Show ke : show 'a) : show (list 'a) =
  let show_it = function
    | Nil -> "Nil"
    | Cons (a, b) -> "Cons (" ^ ke a ^ ", " ^ show_it b ^ ")"
  Show show_it

let implicit functor_list =
  let fmap (f : 'a -> 'b) : list 'a -> list 'b = function
    | Nil -> Nil
    | Cons (x, xs) -> f x :: fmap f xs
  Functor fmap

let implicit applicative_list =
  let ap (f : list ('a -> 'b)) (x : list 'a) : list 'b =
    match f, x with
    | Cons (f, fs), Cons (x, xs) -> f x :: ap fs xs
    | _, _ -> Nil
  let pure (x : 'a) : list 'a = x :: Nil
  Applicative (ap, pure)

let implicit monoid_list =
  let concat (x : list 'a) (y : list 'a) : list 'a =
    match x, y with
    | Nil, y -> y
    | Cons (x, xs), y -> Cons (x, concat xs y)
  Monoid (Nil, concat)

let traverse ?(x : applicative 'f) f = function
  | Nil -> pure Nil
  | Cons (x, xs) -> (::) <$> f x <*> traverse f xs

let foldMapOf l f = getConst & l (Const & f)
let toListOf l = foldMapOf l (fun x -> Cons (x, Nil))

let main =
  let p = (1, "foo")
  print (p |> snd .~ 2)
  let xs = Cons (p, Cons (p, Cons (p, Nil)))
  print (toListOf (traverse & fst) xs)
  print (toListOf (traverse & snd) xs)
  print (xs |> (traverse & snd) .~ 3)
