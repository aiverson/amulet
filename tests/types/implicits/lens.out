tostring : {'a : type}. 'a -> string
writeln : string -> unit
identity : type -> type
Identity : {'a : type}. 'a -> identity 'a
runIdentity : {'gi : type}. identity 'gi -> 'gi
const : {'ha : type}. type -> 'ha -> type
Const : {'a : type}. {'b : 'hn}. 'a -> const 'a 'b
getConst : {'hx : type}. {'b : 'il}. const 'hx 'b -> 'hx
functor : (type -> type) -> type
Functor : {'f : type -> type}. ({'a : type}. {'b : type}. ('a -> 'b) -> 'f 'a -> 'f 'b) -> functor 'f
<$> : {'f : type -> type}. {'a : type}. {'b : type}. functor 'f => ('a -> 'b) -> 'f 'a -> 'f 'b
applicative : (type -> type) -> type
Applicative : {'f : type -> type}. {'ns : type}. ({'a : 'ns}. {'b : 'ns}. 'f ('a -> 'b) -> 'f 'a -> 'f 'b * {'a : type}. 'a -> 'f 'a) -> applicative 'f
<*> : {'f : type -> type}. {'a : type}. {'b : type}. applicative 'f => 'f ('a -> 'b) -> 'f 'a -> 'f 'b
pure : {'f : type -> type}. {'a : type}. applicative 'f => 'a -> 'f 'a
monoid : type -> type
Monoid : {'m : type}. ('m * 'm -> 'm -> 'm) -> monoid 'm
mappend : {'m : type}. monoid 'm => 'm -> 'm -> 'm
mempty : {'uy : type}. monoid 'uy => 'uy
show : type -> type
Show : {'a : type}. ('a -> string) -> show 'a
show : {'a : type}. show 'a => 'a -> string
print : {'xi : type}. show 'xi => 'xi -> unit
show_string : show string
show_int : show int
show_bool : show bool
show_pair : {'a : type}. {'b : type}. show 'a => show 'b => show ('a * 'b)
functor_identity : functor identity
functor_const : {'a : type}. functor (const 'a)
applicative_identity : applicative identity
applicative_const : {'a : type}. monoid 'a => applicative (const 'a)
functor_from_applicative : {'f : type -> type}. applicative 'f => functor 'f
& : {'amu : type}. {'amv : type}. {'ani : type}. ('ani -> 'amv) -> ('amu -> 'ani) -> 'amu -> 'amv
|> : {'aoc : type}. {'aoe : type}. 'aoc -> ('aoc -> 'aoe) -> 'aoe
fst : {'apr : type}. {'f : type -> type}. {'aqe : type}. {'aqh : type}. functor 'f => ('aqh -> 'f 'aqe) -> ('aqh * 'apr) -> 'f ('aqe * 'apr)
snd : {'arz : type}. {'f : type -> type}. {'asn : type}. {'asq : type}. functor 'f => ('asq -> 'f 'asn) -> ('arz * 'asq) -> 'f ('arz * 'asn)
.~ : {'atu : type}. {'amu : type}. {'amv : type}. {'auu : type}. (('auu -> identity 'atu) -> 'amu -> identity 'amv) -> 'atu -> 'amu -> 'amv
^. : {'avt : type}. {'avv : type}. {'b : 'awy}. {'a : type}. {'b : 'axa}. 'avt -> (('a -> const 'a 'b) -> 'avt -> const 'avv 'b) -> 'avv
list : type -> type
Nil : {'a : type}. list 'a
Cons : {'a : type}. ('a * list 'a) -> list 'a
:: : {'ayy : type}. 'ayy -> list 'ayy -> list 'ayy
show_list : {'a : type}. show 'a => show (list 'a)
functor_list : functor list
applicative_list : applicative list
monoid_list : {'a : type}. monoid (list 'a)
traverse : {'f : type -> type}. {'ayy : type}. {'buk : type}. applicative 'f => ('buk -> 'f 'ayy) -> list 'buk -> 'f (list 'ayy)
foldMapOf : {'amu : type}. {'amv : type}. {'b : 'bxk}. {'amu : type}. {'ani : type}. {'b : 'bxn}. (('amu -> const 'ani 'b) -> 'amu -> const 'amv 'b) -> ('amu -> 'ani) -> 'amu -> 'amv
toListOf : {'amu : type}. {'amv : type}. {'b : 'cak}. {'b : 'cal}. {'a : type}. (('a -> const (list 'a) 'b) -> 'amu -> const 'amv 'b) -> 'amu -> 'amv
main : unit
